dll/win32/shell32/dialogs/filetypes.cpp

```diff
@@ -56,204 +158,256 @@ DeleteExt(HWND hwndDlg, LPCWSTR pszExt)
+            Entry->nIconIndex = -IDI_SHELL_EXE;
+        }
+        // Set the icon path to %1 on purpose so PickIconDlg will issue a warning
+        StringCbCopyW(Entry->IconPath, sizeof(Entry->IconPath), IconLocation);
     }
     else
     {
-        DoFileTypeIconLocation(Entry, IconLocation);
+        // Expand the REG_EXPAND_SZ string by environment variables
+        if (ExpandEnvironmentStringsW(IconLocation, Entry->IconPath, _countof(Entry->IconPath)))
+        {
+            Entry->nIconIndex = PathParseIconLocationW(Entry->IconPath);
+            Entry->hIconSmall = DoExtractIcon(Entry->IconPath, Entry->nIconIndex, IconSize, IconSize);
+        }
     }
-
-    return Entry->hIconLarge && Entry->hIconSmall;
+    return Entry->hIconSmall != NULL;
 }
 
 static BOOL
-GetFileTypeIconsByKey(HKEY hKey, PFILE_TYPE_ENTRY Entry)
+GetFileTypeIconsByKey(HKEY hKey, PFILE_TYPE_ENTRY Entry, UINT IconSize)
 {
-    Entry->hIconLarge = Entry->hIconSmall = NULL;
+    Entry->hIconSmall = NULL;
 
-    // Open the "DefaultIcon" registry key
     HKEY hDefIconKey;
     LONG nResult = RegOpenKeyExW(hKey, L"DefaultIcon", 0, KEY_READ, &hDefIconKey);
     if (nResult != ERROR_SUCCESS)
         return FALSE;
 
     // Get the icon location
-    WCHAR szLocation[MAX_PATH + 32] = { 0 };
+    WCHAR szLocation[ICONLOCATION_CCHMAX];
     DWORD dwSize = sizeof(szLocation);
     nResult = RegQueryValueExW(hDefIconKey, NULL, NULL, NULL, LPBYTE(szLocation), &dwSize);
+    szLocation[_countof(szLocation) - 1] = UNICODE_NULL;
 
     RegCloseKey(hDefIconKey);
-
-    if (nResult != ERROR_SUCCESS || szLocation[0] == 0)
+    if (nResult != ERROR_SUCCESS || !szLocation[0])
         return FALSE;
 
-    return GetFileTypeIconsEx(Entry, szLocation);
+    return GetFileTypeIconsEx(Entry, szLocation, IconSize);
+}
+
+static LPCWSTR
+GetProgramPath(PFILE_TYPE_ENTRY p)
+{
+    if (!p->ProgramPath[1] && !p->ProgramPath[0])
+    {
+        DWORD cch = _countof(p->ProgramPath);
+        if (FAILED(AssocQueryStringW(ASSOCF_INIT_IGNOREUNKNOWN, ASSOCSTR_EXECUTABLE,
+                                     p->FileExtension, NULL, p->ProgramPath, &cch)))
+        {
+            p->ProgramPath[0] = UNICODE_NULL;
+            p->ProgramPath[1] = TRUE;
+        }
+    }
+    return p->ProgramPath;
 }
 
 static BOOL
 QueryFileDescription(LPCWSTR ProgramPath, LPWSTR pszName, INT cchName)
 {
-    SHFILEINFOW FileInfo = { 0 };
-    if (SHGetFileInfoW(ProgramPath, 0, &FileInfo, sizeof(FileInfo), SHGFI_DISPLAYNAME))
+    SHFILEINFOW fi;
+    fi.szDisplayName[0] = UNICODE_NULL;
+    if (SHGetFileInfoW(ProgramPath, 0, &fi, sizeof(fi), SHGFI_DISPLAYNAME))
     {
-        StringCchCopyW(pszName, cchName, FileInfo.szDisplayName);
+        StringCchCopyW(pszName, cchName, fi.szDisplayName);
         return TRUE;
     }
-
     return !!GetFileTitleW(ProgramPath, pszName, cchName);
 }
 
+static LPCWSTR
+GetAppName(PFILE_TYPE_ENTRY p)
```

Comment:
Same here.

Reply:
Confusing? The entire function is 10 lines of code.